core/state.py

```diff
@@ -0,0 +1,88 @@
+from collections import defaultdict
+from functools import wraps
+import threading
+
+if False:
+    from typing import Callable, DefaultDict, Dict, List, Optional, TypeVar
+    from mypy_extensions import TypedDict
+    from .git_mixins.status import FileStatus
+
+    T = TypeVar('T')
+
+    RepoPath = str
+    RepoStatus = TypedDict(
+        'RepoStatus',
+        {
+            'detached': bool,
+            'branch': Optional[str],
+            'remote': Optional[str],
+            'clean': bool,
+            'ahead': Optional[int],
+            'behind': Optional[int],
+            'gone': bool,
+            'rebasing': bool,
+            'rebase_branch_name': Optional[str],
+            'merging': bool,
+            'merge_head': Optional[str],
+            'file_statuses': List[FileStatus],
+            'staged': List[FileStatus],
+            'unstaged': List[FileStatus],
+            'untracked': List[FileStatus],
+            'conflicts': List[FileStatus],
+            'short_status': str,
```

Comment:
'staged/unstaged/untracked/conflicts' is derived state from 'file_statuses'. 'short_status' is derived from all the other properties. It is usually better to *not* store derived values in the main store.